pandas/tests/indexes/test_numeric.py

```diff
@@ -1206,3 +1213,10 @@ def test_join_outer(self):
         tm.assert_index_equal(res, eres)
         tm.assert_numpy_array_equal(lidx, elidx)
         tm.assert_numpy_array_equal(ridx, eridx)
+
+    def test_insert(self):
+        # GH 18295 (test missing)
+        expected = UInt64Index([0, 0, 1, 2, 3, 4])
+        for na in (np.nan, pd.NaT, None):
```

Comment:
hmm, I *think* this should coerce to a ``FloatIndex``, cc @gfyoung

Reply:
This was my intuition as well, and how I originally wrote the test (flowed through from `Numeric` class).  The logic behind `expected` is based on the `_na_value` of `UInt64Index` being 0:

```python
In [2]: pd.UInt64Index._na_value
Out[2]: 0
```
So, it'd be a matter of altering that to `np.nan` if we want it to coerce to `Float64Index`.

Reply:
hmm, this should mimic ``Int64Index``, in that it *cannot* hold na, so yes pls change this logic. (possibly this might break other things, not sure). may need to do this as a pre-curser (or you do it post as well, lmk).

Reply:
Tested out changing the default value to `np.nan`, and it produced the expected result of coercing to `Float64Index`.  It broke two tests though, both involving `Index.where`, so I didn't include it in the most recent commit.  The tests that broke are:

1) `test_where_array_like` fails due to `expected` raising a `ValueError`:
https://github.com/pandas-dev/pandas/blob/5c0b20a3a383e69278f6347a958ccb1388d0334e/pandas/tests/indexes/test_numeric.py#L742-L748
The change makes `_nan` -> `np.nan`, so this essentially becomes `pd.Index([np.nan, 1, 2, 3, 4], dtype='uint64')`, which raises `ValueError:  cannot convert float NaN to integer`.  This seems like the expected behavior.  Or should this coerce to `Float64Index`?

2) `test_where` fails because `result` gets coerced to `Int64Index`:
https://github.com/pandas-dev/pandas/blob/5c0b20a3a383e69278f6347a958ccb1388d0334e/pandas/tests/indexes/test_numeric.py#L729-L733
This actually happens in 0.21.0 too, and is only now apparent due to testing the change.  Essentially, if any non-uint64 value gets passed as other (in this case `np.nan` due to the change), the index gets coerced, even if the mask results in no changes.  Reproducing this error on 0.21.0:

```python
In [2]: idx = pd.UInt64Index(range(5))
   ...: idx
   ...:
Out[2]: UInt64Index([0, 1, 2, 3, 4], dtype='uint64')

In [3]: idx.where(idx < 100, np.nan)
Out[3]: Int64Index([0, 1, 2, 3, 4], dtype='int64')
```
So far, I've traced the coercing back to `Index` not properly casting to `uint64`:
```python
In [4]: pd.Index([0., 1., 2., 3., 4.], dtype='uint64')
Out[4]: Int64Index([0, 1, 2, 3, 4], dtype='int64')
```
Works fine if ints are passed instead of floats:
```python
In [5]: pd.Index([0, 1, 2, 3, 4], dtype='uint64')
Out[5]: UInt64Index([0, 1, 2, 3, 4], dtype='uint64')
```
Haven't dug all the way into this yet, but am leaning towards making a separate issue/PR, unless the fix ends up being trivial.

Reply:
is this fixed here? or the other issue fix should be merged first?

Reply:
Will open a separate PR that should be merged first since there are a couple unrelated fixes.  My bad, said that in the original comment above, but then edited over it!

Reply:
np; just reference this pr from new one